src/draw/shapes.typ

```diff
@@ -705,16 +725,60 @@
         offset-anchor: anchor,
         transform: ctx.transform
       )
-      
-      let style = styles.resolve(ctx.style, style, root: "rect")
-      let (x1, y1, z1) = a
-      let (x2, y2, z2) = b
-      let drawables = drawable.path(
-        path-util.line-segment(((x1, y1, z1), (x2, y1, z2), (x2, y2, z2), (x1, y2, z1))),
-        fill: style.fill,
-        stroke: style.stroke,
-        close: true,
-      )
+
+      let drawables = {
+        let z = z1
+        let get-corner-pts(radius, pt, a, b) = {
+          let (rx, ry) = radius
+          if rx > 0 or ry > 0 {
+            let (xa, ya) = a
+            let (xb, yb) = b
+            (vector.add(pt, (xa * rx, ya * ry)),
+             vector.add(pt, (xb * rx, yb * ry)))
+          } else {
+            (pt, pt)
+          }
+        }
+
+        // Get segments for arc between start- and stop angle, starting
+        // at point.
+        let corner-arc(radius, start-angle, stop-angle, pt) = {
+          let (rx, ry) = radius
+          if rx > 0 or ry > 0 {
+            let (x, y, ..) = pt
+            drawable.arc(x, y, z, start-angle, stop-angle, rx, ry).segments
+          }
+        }
+
+        // Compute all eight corner points:
+        //
+        //    p1-------p2
+        //   / |       | \
+        // p0--+       +--p3
+        //  |             |
+        // p7--+       +--p4
+        //   \ |       | /
+        //    p6-------p5
+        //
+        // If a corner has radius (0,0), both of its
+        // corner points are the same.
+        let (p0, p1) = get-corner-pts(nw, (x1, y2, z), ( 0,-1), ( 1, 0))
+        let (p2, p3) = get-corner-pts(ne, (x2, y2, z), (-1, 0), ( 0,-1))
+        let (p4, p5) = get-corner-pts(se, (x2, y1, z), ( 0, 1), (-1, 0))
+        let (p6, p7) = get-corner-pts(sw, (x1, y1, z), ( 1, 0), ( 0, 1))
+
+        let segments = ()
+        segments += corner-arc(nw, 180deg, 90deg, p0)
+        segments += (path-util.line-segment((p1, p2)),)
+        segments += corner-arc(ne, 90deg, 0deg, p2)
+        segments += (path-util.line-segment((p3, p4)),)
+        segments += corner-arc(se, 0deg, -90deg, p4)
+        segments += (path-util.line-segment((p5, p6)),)
+        segments += corner-arc(sw, -90deg, -180deg, p6)
+        segments += (path-util.line-segment((p7, p0)),)
+
+        drawable.path(segments, fill: style.fill, stroke: style.stroke, close: true)
+      }
```

Comment:
I don't understand how this works and it scares me... Could you tidy it up a bit and some more detailed comments for each step?